Tee object can now store received error records to a variable.

JamesWTruher opened the pull request:
This addresses an aspect of issue #5184 

<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This adds the parameter `-ErrorRecordVariable` (Alias `ERV`) to `Tee-Object`
if `Tee-Object` receives an ErrorRecord in the input stream and `-ErrorRecordVariable` is included,
`Tee-Object` will assign those error records to the variable. This enables the following scenario:

```powershell
PS> Get-Item $PSHOME,doesnotexist 2>&1 | Tee-Object -var out -ERV ee

    Directory: /Users/james/src/github/forks/jameswtruher/PowerShell-1/src/powershell-unix/bin/Debug/net9.0/osx-x64

UnixMode         User Group         LastWriteTime         Size Name
--------         ---- -----         -------------         ---- ----
drwxr-xr-x      james staff        5/1/2024 11:34         9024 publish

PS> $out

    Directory: /Users/james/src/github/forks/jameswtruher/PowerShell-1/src/powershell-unix/bin/Debug/net9.0/osx-x64

UnixMode         User Group         LastWriteTime         Size Name
--------         ---- -----         -------------         ---- ----
drwxr-xr-x      james staff        5/1/2024 11:34         9024 publish

PS> $ee
Get-Item: Cannot find path '/Users/james/src/github/forks/jameswtruher/PowerShell-1/src/Microsoft.PowerShell.Commands.Utility/commands/utility/doesnotexist' because it does not exist.
PS>
```

It works with file, literalfile, and variable parameter sets.
This works with native executables as well, as when stderr is written, we create an error record.
Errors created by `Tee-Object` are emitted in the normal way, it is only the input object which is inspected and handled by this new behavior.
Note: the `2>&1` is required to ensure that the error stream is emitted into the output stream

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

JamesWTruher committed:
Address an aspect of issue #5184.

Tee object can now store error records to a variable.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Tee-Object.Tests.ps1:
```diff
@@ -92,4 +92,18 @@ Describe "Tee-Object DRT Unit Tests" -Tags "CI" {
     }
 
+    It "Positive Variable/ErrorRecordVariable Test" {
+        $results = Get-Item $PSHOME,doesnotexist1,$PWD,doesnotexist2 2>&1 | Tee-Object -Variable outputVar -ErrorRecordVariable errVar
```
I think you should have a test case with a native command to ensure that use case doesn't break

JamesWTruher committed:
Fix CodeFactor issues.

Add a native executable test.

SteveL-MSFT reviewed (approved):
LGTM

Jaykul commented:
From the example, you seem to have _eaten_ the error!?

Tee-Object should still pass through errors, surely.

Jaykul commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Tee-Object.Tests.ps1:
```diff
@@ -92,4 +92,25 @@ Describe "Tee-Object DRT Unit Tests" -Tags "CI" {
+    }
+
+    It "Positive File/ErrorRecordVariable Test" {
+        $results = Get-Item $PSHOME,doesnotexist1,$PWD,doesnotexist2 2>&1 | Tee-Object -File $tempFile -ErrorRecordVariable errVar
```
Can we add a test where you Tee _a second time_ to prove the error is still passed through?

Jaykul commented:
Incidentally, is it **even possible** to get output in `-OutVariable` or `-ErrorVariable` from this command?

I mean, even when I do, say `| Tee-Object -File Variable:\Foo` or `| Tee -File NoSuchDrive:\log.txt` I do not get an error.

It's weird to me that it has _two parameters_ for output and error variables, but the one we all **EXPECT** ... does not work. It's pretty confusing.

SteveL-MSFT commented:
> Incidentally, is it **even possible** to get output in `-OutVariable` or `-ErrorVariable` from this command?
> 
> I mean, even when I do, say `| Tee-Object -File Variable:\Foo` or `| Tee -File NoSuchDrive:\log.txt` I do not get an error.
> 
> It's weird to me that it has _two parameters_ for output and error variables, but the one we all **EXPECT** ... does not work. It's pretty confusing.

If you redirect `stderr` to `stdout` and pipe to `Tee-Object`, you'll get both into the output stream and can do whatever you want.  We thought the ask was to be able to handle `stderr` separately thus the new variable to de-multiplex it.  Certainly you could do this yourself just by looking at the type of object and this was a convenience feature.

mklement0 commented:
@SteveL-MSFT, while I can appreciate what this PR is trying to achieve, I think it is ultimately the wrong way to go about it, especially given that it _requires modifying the call that provides the input_ (`2>&1`) in order to work meaningfully.

* #20381 is a great extension to how `>` and `>>` work, but it is currently limited to _consuming_ the targeted stream; that is, unlike with the common `-*Variable` parameters, the stream isn't passed through.
 
* Extending the `> variable:foo` syntax to optionally support pass-through - [as previously discussed](https://github.com/PowerShell/PowerShell/issues/5184#issuecomment-343741446), where you yourself suggested something like `>| variable:foo` - strikes me as the better alternative to extending `Tee-Object`.

SteveL-MSFT commented:
@mklement0 at this time we're not entertaining changes to the pipe syntax `2>|`, for example, so within the confines of the current syntax both this change and the redirect to variable change should support the requested scenarios.  I would agree that `2>&1 | Tee-Object` isn't obvious, but it's supported today and this is just an extension of that

mklement0 commented:
@SteveL-MSFT: Note that, as currently described, this feature seems to provide no additional functionality beyond the following.

```powershell
$out = Get-Item $PSHOME, doesnotexist 2>variable:ee
```

Given that, along with:

* the obscurity of requiring `2>&1` in the upstream pipeline segment

* the departure from the `Tee-Object`'s core mandate - to pass data _through_ - which, from what it seems, won't be done for the `ErrorRecord` objects in the pipeline (a concern expressed earlier by @Jaykul's).

this still strikes me as not worthwhile - even if you're not willing to tackle something like `2>|`

JamesWTruher committed:
Remove erroneous return when storing an error record.

It should go into the results but _also_ go to the error variable.
fi test for same.

JamesWTruher committed:
fix improper case of testexe.

mklement0 commented:
@SteveL-MSFT, assuming you still see value in this PR (if so, please let us know _why_),
please consider the "collateral damage" relative to the perceived benefit:

* Increased implementation complexity of `Tee-Object`
* Documentation effort.
* Increased cognitive load on users who need to scan the more complex command-line help / documentation, even though they may never avail themselves of this feature.
